### ISSUE TITLE
Implement TimScheme as a language running on the Guile VM

### ISSUE THREAD

**Issue by mardukbp** [texmacs/TimScheme#2]
There are many things I would like to implement or improve in TeXmacs and some of them require Guile 2. For example, using the PEG parser one could easily implement syntax highlighting for many languages, without having to recompile TeXmacs. Therefore I am very interested in upgrading TeXmacs to support Guile 2.

However, I have not found a clear and detailed explanation of what prevents it from happening. From the existence of this repo and also after reading some Scheme files from the TeXmacs source I gather that TeXmacs mostly implements its own DSL, which you have called TimScheme. I suppose that the way this was done is no longer supported by Guile 2.

Reading the Guile Manual I learned that it [is possible](https://www.gnu.org/software/guile/manual/guile.html#Other-Languages) to implement a language running on top of the Guile VM. ECMAScript and Emacs Lisp are already included and Lua has been recently implemented too. Maybe it is a crazy idea, but I was thinking that it should be possible to implement TimScheme as a language running on the Guile VM. That way, TeXmacs would continue to run in future versions of Guile (Guile 3 is on the works).

**Comment by da-liii** [comment]
https://www.gnu.org/software/guile/manual/guile.html#Compiling-to-the-Virtual-Machine

Great

**Comment by da-liii** [comment]
@mardukbp  https://github.com/texmacs/guile

**Comment by da-liii** [comment]
@mardukbp  Let's write down the spec here.

**Comment by mgubi** [comment]
I've made progress in porting to guile2. TeXmacs works almost correclty. Some issues remains (essentially linked to compilation and to encoding of strings) but now seems less complicated than it was last year. Please check out the guile2 branch in my repository: https://github.com/mgubi/texmacs/tree/guile2

**Comment by mgubi** [comment]
The issues are:

1. TeXmacs strings are Cork encoded while Scheme strings can contains glyphs which cannot be encoded in Cork. So I'm not sure what to do there. In general is not very clear to me how the whole issue of encodings is to be treated at boundaries like Texmacs/Qt and TeXmacs/Guile.

2. Guile 2 has separated expansion and compilation phases, this causes some problem when macros need functions at expansion time. This means that eval-when forms has to be used to indicate to the expander to evaluate some forms during expansion. In general TeXmacs' scheme code could be cleaned up and functions only needed in  macro definitions put it the local context of the macro.

**Comment by mardukbp** [comment]
@mgubi This is great news! Thanks. Are you aware that Guile 3 is almost there?

**Comment by mgubi** [comment]
@mardukbp Yes, I'm testing also on guile 2.9 (which will be Guile 3, as soon as it is released). Seems also fine. There is essentially no difference on  TeXmacs's side between Guile2 and Guile3.

**Comment by da-liii** [comment]
@mgubi 

Any guide to try the guile 2 branch ?

I'd like to help.

**Comment by da-liii** [comment]
Well, I am using Debian sid.

I have to solve this bug first: https://savannah.gnu.org/bugs/?55986

**Comment by mgubi** [comment]
configure is not working fine with Guile 2.2. So What i do is to configure with Guile 1.8, and then manually tune the produced Makefile, src/makefile and config.h to point to the right libraries and include file. The symbol GUILE_D has to be defined and GUILE_C undefined in config.h. After this TeXmacs should compile. You have to run with GUILE_AUTO_COMPILE=0 set in the environment. If you want to test the compilation then  GUILE_AUTO_COMPILE=1. But recall that once modules are compiled they will override the uncompiled ones. So if you want to go back to interpreted code you have to cancel the cache in ~/.cache/guile/(whatever directory TeXmacs is in)

**Comment by mgubi** [comment]
The branch guile2-compile contains specifics changes for compilation, mostly eval-when in appropriate places.

**Comment by mgubi** [comment]
You can also use the development branch of guile.

**Comment by da-liii** [comment]
OK, I may write some code in `CMakeLists.txt` .

Autotools is a nightmare for me.

**Comment by da-liii** [comment]
I planned to extract the minimum code for texmacs kernel and guile scheme kernel in this repo.
And then try to upgrade guile 1.8 to guile 2.x .

I do think it is a good method. Especially for Guile exports who are not familiar with the code base of GNU TeXmacs.

I will continue my work in this repo later.

**Comment by da-liii** [comment]
@mgubi   Really cool. I just build it using CMake without any modification on your guile2 branch.

It works fine.

We may begin our guile 2 migration after the release of TeXmacs 2.1 .

**Comment by da-liii** [comment]
@mgubi  Your former CMake definition is realy solid. I haven't changed the guile part. It works fine. I'm surprised. :)

**Comment by mgubi** [comment]
@sadhen I think others worked on the CMake definitions after I sketched the first implementation. Not sure who however. I'm not using them, so I'm also surprised it works. :) It would make sense have some build system based on Guile (like Guix has) since anyway we have to have it installed on the machine. Maybe it is worth to give a look at the possibilities.

**Comment by mgubi** [comment]
@sadhen The guile2 branch is still far from ready, it somewhat works only in non-compiled mode and this is not good since the performances then could be worser than with Guile 1.8 (which had a very performing interpreter written in C, while the interpreter in 2+ is written in Scheme and not so fast). There are problems with compiled modules (they compile but more subtle bugs appears), and with encoding of some strings. So I think we have to keep the code in a developing branch and not merge it until we are not sure that is working ok and that we do not have performance regressions.

**Comment by mgubi** [comment]
@sadhen btw, in the qt5 branch you find also my modifications for Qt5, which now seems to work fine. On the Mac one needs a patch which I submitted to Qt and will be integrated in the next release of 5.12. You can get the patched sources from the devel branch at Qt. Only qtbase is needed by TeXmacs.

**Comment by mgubi** [comment]
Wrt. to scheme: it would be however nice to have the guile-dependent code well separated by the main codebase, I still have some curiosity to explore other possibilites for the scheme side. We cannot really use all the power of Guile, since for example, I'm not really sure we cannot reliably use continuations since we call scheme from C++ code. A cool project could also be to have TeXmacs as a library for Guile and not the other way around, that is go into C++ code only for computational intensive tasks and keep all the UI code in Scheme.

**Comment by mgubi** [comment]
Lot of cool things to think about...
